keras/src/distribution/tensor_parallel/autoconfig.py

```diff
@@ -0,0 +1,241 @@
+import functools
+
+from keras.src import layers
+from keras.src.backend import distribution_lib
+from keras.src.distribution.tensor_parallel.tensor_layout import LayoutMap
+from keras.src.distribution.tensor_parallel.tensor_layout import (
+    split_tensor_for_parallelism,
+)
+
+
+def analyze_dense_layer(layer):
+    """
+    Classifies a Dense layer based on its input/output dimensions.
+
+    This function determines if a Dense layer represents an 'up_projection'
+    (expansion) or a 'down_projection' (contraction) based on a heuristic
+    threshold. This classification dictates how the weights are sharded.
+
+    Heuristic:
+        - Expansion: Output dimension > (Input dimension * 1.5)
+        - Contraction: Input dimension > (Output dimension * 1.5)
+
+    Args:
+        layer (keras.layers.Layer): The layer instance to analyze.
+
+    Returns:
+        str: One of 'up_projection', 'down_projection', or 'dense'.
+    """
+    if not isinstance(layer, layers.Dense):
+        return "dense"
+
+    input_dim = None
+    output_dim = None
+
+    if hasattr(layer, "_kernel") and layer._kernel is not None:
+        kernel_shape = layer._kernel.shape
+        if len(kernel_shape) == 2:
+            input_dim = kernel_shape[0]
+            output_dim = kernel_shape[1]
+    elif hasattr(layer, "kernel") and layer.kernel is not None:
+        kernel_shape = layer.kernel.shape
+        if len(kernel_shape) == 2:
+            input_dim = kernel_shape[0]
+            output_dim = kernel_shape[1]
+
+    if input_dim is None or output_dim is None:
+        if hasattr(layer, "units"):
+            output_dim = layer.units
+        else:
+            return "dense"
+
+        if (
+            hasattr(layer, "input_shape")
+            and layer.input_shape
+            and len(layer.input_shape) > 1
+        ):
+            input_dim = layer.input_shape[-1]
+        else:
+            return "dense"
```

Comment:
Shouldn't this be done only if `input_dim` is None? (as opposed to if either `input_dim` or `output_dim` is None)